### ISSUE TITLE
annotations of material or course according to (Bio)schemas specification

### ISSUE THREAD

**Issue by abotzki** [LiaScript/LiaScript#118]
Hi,

as discussed some time ago, I send you two example of an annotation of training material (and courses = evetns) which we use:

[example Learning material](https://validator.schema.org/#url=https%3A%2F%2Ftess.elixir-europe.org%2Fmaterials%2Fbulk-rnaseq-from-counts-to-differential-expression%20)

Following this specification, the TeSS team extracts relevant data from the registry system that can be transferred to the JSON-LD object in the HTML header of the individual material page.

In an ideal scenario, the original material page has also an annotation.

For events, we use Course/CourseInstance: see [example Course/CourseInstance](https://validator.schema.org/#url=https%3A%2F%2Ftraining.vib.be%2Fall-trainings%2Fanalysis-single-cell-rnaseq-data-10x-genomics-leuven-4)

The doc: https://bioschemas.org/profiles/

Please reach out if you have some question on this matter.

**Comment by andre-dietrich** [comment]
@abotzki Thanks a lot, I will try this out and ask my local chat-gpt and @SebastianZug for help and come back to you ... I will try to add basic functionality into our exporter, that shall be afterwards linked to a course document .... my current tool of choice for handling linked data is comunica of would you suggest another one ;-)

**Comment by abotzki** [comment]
Comunica looks very powerful. I will need to study a little bit more what
you could use it for.
From what I understand you could query annotated resources quite easily.

For the time being, we are generating the JSON-LD objects by extracting
information from a Drupal CMS which is the technical backend of our course
registration system.

Via a sitemap.xml, these JSON-LD objects are provided to a bespoke scraper
which consumes the information and puts it into the TeSS database.

Essentially methods which work fine in our small ELIXIR ecosystem. At the
same time, we provide annotated data for google etc to pick up.

**Comment by andre-dietrich** [comment]
Hi @abotzki , for testing-purposes I added a new export format to the exporter-tool:

https://www.npmjs.com/package/@liascript/exporter#rdf--json-ld

It generates a simple jsonld-file based on the LiaScript meta-information ... I used Course instead of EducationalResource, but this can be fixed via `--rdf-type`. 

At the moment I am not sure how to add additional information, maybe this could be included into the header-comment or by adding a special command-option to the exporter, that reads in a template-jsonld and adds the data from the Readme. What do you think?

You can also try out the exporter with URLs as `--input | -i` parameter and `--rdf-preview` ...

**Comment by abotzki** [comment]
Hi @andre-dietrich thanks for all this work. I did not yet run it but I will soon.

Just for inspiration how e.g. the Galaxy training network approaches this:

Have a look at this training material:
https://raw.githubusercontent.com/galaxyproject/training-material/main/topics/single-cell/tutorials/scrna-raceid/tutorial.md

In the md file is also a yml header with some of the relevant information.

Apart from this, the information for the JSON-LD object for each course or learningResource s extracted and merged from various sources and put into the header of the rendered HTML page.

see code here:
https://github.com/galaxyproject/training-material/blob/main/_plugins/jekyll-jsonld.rb

So, obviously the header-comment is an option as well as the suggestion to merge on export with other information from a template-jsonld file.

Importantly, I was wondering whether you could make this JSON-LD objects also part of the HTML header of the individual material next to have it separately stored?

**Comment by andre-dietrich** [comment]
Yes, that could be added to a web and project output generated by the exporter tool ... I will have a look at the example, thank you for that ... The `--rdf-template` option is locally included, I need to run a few experiments, if this works like expected and then publish it ...

**Comment by andre-dietrich** [comment]
@abotzki ... Hi, I just published the update with the --rdf-template option ... with this you can us a file or http-reference of a json-ld file, only fields that are not defined within the template will be added by the exporter. `keywords` are an exception to this rule and will be added, if they have been defined within the template.

However, I wanted to generate also a more complex json and add this to the project-export

https://tubaf-ifi-liascript.github.io

You can see the output in action here:

https://tubaf-ifi-liascript.github.io/

this will require a grouping of courses on two levels, since it is also possible to generate sub-collections, that are presented their own box with smaller cards.

However, I am at the moment not sure, which format fits best. I think of ItemList, which may contain another ItemList for sub-collections of courses... Or, do you know or maybe use another structure?

**Comment by abotzki** [comment]
@andre-dietrich Thanks a lot for your work. I'll try it out soon. 

Regarding the overview of the project export an dhow you organise this is an open question in our community as well.

We tend to think about learning paths as well as collections/sub-collections. You can imagine that this is not so easy to get the head around it and keep it simple in a presentation format. 

For the time being, we have the concept of a simple collection which is just a list. 

But in our future activitites within ELIXIR, we would like to include learning paths as well. This will be more difficult since you are linking "modules" (e.g. items of the list) with each other.
For instance, the LearningResource definition has a hasPart or isPartOf property which is an initial starting point to create links between objects of LearningResources. How this overall picture looks is not yet discussed.

**Comment by abotzki** [comment]
I just found this tool: https://linkml.io/linkml/ - not sure whether is useful compared to comunica.

**Comment by abotzki** [comment]
@andre-dietrich We did not yet discuss the querying of multiple educational resources. In contrast, I like you using ItemList for the collection of courses. 
It has been included in the discussions in the ELIXIR All Hands meeting which takes place beginning of June 2023.

**Comment by abotzki** [comment]
@andre-dietrich I see that for collections of learning materials this definition instead of an ItemList will most probably be selected: [https://schema.org/Collection](https://schema.org/Collection) - still to be confirmed. Furthermore, discussions about learning paths will be started and this will take time.

**Comment by andre-dietrich** [comment]
@abotzki thanks for the info, I will have a look at it ...
